File: 6_binary_trees/C++/Concepts/binary_tree.cpp
#include <iostream>
using namespace std;

// THIS IS BINARY SEARCH TREE

// insertion
// find
// delete
class Node{
    public:
    int data;
    Node* left;
    Node* right;
    Node(int data){
        this->data = data;
        this->left = NULL;
        this->right = NULL;
    }
};

class BST{
    public:
    Node* root;
    BST(int root_val){
        root = new Node(root_val);
    }

    void insert(int val){
        Node* newnode = new Node(val);
        if(root == NULL){
            root = newnode;
            return;
        }
        Node* current = root;
        while(true){
            if (val < current->data){
                if(current->left == NULL){
                    current->left = newnode;
                    break;
                }
                current = current->left;
            }
            else{
                if(current->right == NULL){
                    current->right = newnode;
                    break;
                }
                current = current->right;
            }
        }
    }

    bool find(int val){
        Node* current = root;
        while(current != NULL){
            if (val < current->data){
                current = current->left;
            }
            else if(val > current->data){
                current = current->right;
            }
            else{
                return true;
            }
        }
        return false;
    }
};

int main(){

    BST bst = BST(7);
    bst.insert(6);
    bst.insert(9);
    bst.insert(4);

    cout << bst.find(9) << "\n";

    return 0;
}

File: 6_binary_trees/Python/Concepts/binary_tree.py
class Node:
    def __init__(self, value):
        self.value = value
        self.left = None
        self.right = None

class BST:
    def __init__(self, root_val):
        self.root = Node(root_val)
    
    def insert(self, value):
        newnode = Node(value)
        if self.root==None:
            self.root = newnode
            return
        
        current = self.root
        while True:
            if value < current.value:
                if current.left == None:
                    current.left = newnode
                    break
                current = current.left
            else:
                if current.right == None:
                    current.right = newnode
                    break
                current = current.right

    def find(self, value):
        current = self.root
        while(current != None):
            if value < current.value:
                current = current.left
            elif value > current.value:
                current = current.right
            else:
                return True
        return False


if __name__ == "__main__":
    bst = BST(7)
    bst.insert(10)
    bst.insert(3)
    bst.insert(5)

    print(bst.root.value)
    print(bst.root.left.value)
    print(bst.root.right.value)
    print(bst.root.left.right.value)

    print(bst.find(5))
    print(bst.find(23))